opentelemetry-sdk/src/metrics/instrument.rs

```diff
@@ -207,27 +212,53 @@ impl StreamBuilder {
     ///
     /// A Result containing the new Stream instance or an error if the build failed.
     pub fn build(self) -> Result<Stream, Box<dyn Error>> {
-        // TODO: Add same validation as already done while
-        // creating instruments. It is better to move validation logic
-        // to a common helper and call it from both places.
-        // The current implementations does a basic validation
-        // only to close the overall API design.
+        // TODO: Avoid copying the validation logic from meter.rs,
+        // and instead move it to a common place and do it once.
+        // It is a bug that validations are done in meter.rs
+        // as it'll not allow users to fix instrumentation mistakes
+        // using views.
 
-        // if name is provided, it must not be empty
+        // Validate name if provided
         if let Some(name) = &self.name {
             if name.is_empty() {
-                return Err("Stream name must not be empty".into());
+                return Err(INSTRUMENT_NAME_EMPTY.into());
+            }
+
+            if name.len() > super::meter::INSTRUMENT_NAME_MAX_LENGTH {
+                return Err(INSTRUMENT_NAME_LENGTH.into());
+            }
+
+            if name.starts_with(|c: char| !c.is_ascii_alphabetic()) {
+                return Err(INSTRUMENT_NAME_FIRST_ALPHABETIC.into());
+            }
+
+            if name.contains(|c: char| {
+                !c.is_ascii_alphanumeric()
+                    && !super::meter::INSTRUMENT_NAME_ALLOWED_NON_ALPHANUMERIC_CHARS.contains(&c)
+            }) {
+                return Err(INSTRUMENT_NAME_INVALID_CHAR.into());
             }
         }
 
-        // if cardinality limit is provided, it must be greater than 0
+        // Validate unit if provided
+        if let Some(unit) = &self.unit {
+            if unit.len() > super::meter::INSTRUMENT_UNIT_NAME_MAX_LENGTH {
+                return Err(INSTRUMENT_UNIT_LENGTH.into());
+            }
+
+            if unit.contains(|c: char| !c.is_ascii()) {
+                return Err(INSTRUMENT_UNIT_INVALID_CHAR.into());
+            }
+        }
+
+        // Validate cardinality limit
         if let Some(limit) = self.cardinality_limit {
             if limit == 0 {
                 return Err("Cardinality limit must be greater than 0".into());
             }
         }
 
-        // If the aggregation is set to ExplicitBucketHistogram, validate the bucket boundaries.
+        // Validate bucket boundaries if using ExplicitBucketHistogram
```

Comment:
Now that we are doing the validation while building the stream, we could get rid of this code:

https://github.com/open-telemetry/opentelemetry-rust/blob/8c29ca7e21cf0c041695a50d683504f9cbeca515/opentelemetry-sdk/src/metrics/internal/histogram.rs#L95-L100

Reply:
good point, will tackle it soon.